platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/CompoundAssignmentResult.java

```diff
@@ -0,0 +1,90 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import com.google.common.base.Preconditions;
+import org.gradle.api.internal.provider.support.SupportsCompoundAssignment;
+import org.gradle.api.provider.Provider;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * A helper class to implement an intermediate result of a compound assignment operation, like "+=".
+ * It is then assigned to the left-hand side operand. When the LHS is a Property-typed property of some Gradle-enhanced object, then the assignment action is invoked.
+ */
+public final class CompoundAssignmentResult<T> extends AbstractMinimalProvider<T> implements SupportsCompoundAssignment.Result<Provider<T>> {
+    private final ProviderInternal<T> value;
+    private final Object owner;
+    @Nullable
+    private Runnable assignToOwnerAction;
+
+    /**
+     * Creates the result for {@code owner <OP> rhs} operation.
+     *
+     * @param value the intermediate value of the operation, used when LHS is a variable or non-Gradle enhanced property
+     * @param owner the LHS operand of the compound operation
+     * @param assignToOwnerAction the mutation of the owner
+     */
+    public CompoundAssignmentResult(ProviderInternal<T> value, Object owner, Runnable assignToOwnerAction) {
+        this.value = value;
+        this.owner = owner;
+        this.assignToOwnerAction = Objects.requireNonNull(assignToOwnerAction);
+    }
+
+    public boolean isOwnedBy(Object target) {
+        // It might be that this object was unwrapped, in which case it is considered a normal provider.
+        return assignToOwnerAction != null && target == owner;
+    }
+
+    public void assignToOwner() {
+        Runnable action = assignToOwnerAction;
+        Preconditions.checkState(action != null, "The property is already consumed by the owner");
+        assignToOwnerAction = null;
+        action.run();
+    }
+
+    @Override
+    protected Value<? extends T> calculateOwnValue(ValueConsumer consumer) {
+        return value.calculateValue(consumer);
+    }
+
+    @Override
+    public boolean calculatePresence(ValueConsumer consumer) {
+        return value.calculatePresence(consumer);
+    }
+
+    @Override
+    public ExecutionTimeValue<? extends T> calculateExecutionTimeValue() {
+        return value.calculateExecutionTimeValue();
+    }
+
+    @Nullable
+    @Override
+    public Class<T> getType() {
+        return value.getType();
+    }
+
+    @Override
+    @Nullable
+    public Provider<T> unwrap() {
+        // When the expression involves a variable on the left side as opposed to a filed, then this provider becomes its value.
+        // It must lose all "magical" properties towards its owner, because it may be used to set its value outside the original expression.
+        assignToOwnerAction = null;
+        return null;
```

Comment:
❓ The PR description explains that it's non-trivial to return a workable Provider from here, but what other alternatives do we have?

I wonder if returning null can result in some hard to debug problems if users do use the compound assignment as an expression. Maybe not intentionally, but as a result of the PAPI migration.

Are there any options to blow up instead, if the resulting value is used in any way? We can also supply an actionable error message, that such code must be rewritten, instead of silently returning null, which maybe be ignored.

Reply:
> The PR description explains that it's non-trivial to return a workable Provider from here, but what other alternatives do we have?

There is an internal spec for that: https://docs.google.com/document/d/1J7xwHrP4vk7Mtt3GsXgaKHL09EOZ-EqwxdKS8AE6dIo/edit?tab=t.0#heading=h.x5xd0ggtdd90

A dummy value adhering to Groovy conventions can be somewhat better option, though there is always a chance of hitting the issue with it pretty far away from the original expression.

I suggest we revisit it in a separate PR to keep this one shorter.